libcudacxx/include/cuda/std/__cccl/assert.h

```diff
@@ -0,0 +1,123 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __CCCL_ASSERT_H
+#define __CCCL_ASSERT_H
+
+#include <cuda/std/__cccl/compiler.h>
+#include <cuda/std/__cccl/system_header.h>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__cccl/diagnostic.h>
+
+#if defined(_DEBUG) || defined(DEBUG)
+#  ifndef _CCCL_ENABLE_DEBUG_MODE
+#    define _CCCL_ENABLE_DEBUG_MODE
+#  endif // !_CCCL_ENABLE_DEBUG_MODE
+#endif // _DEBUG || DEBUG
+
+// Automatically enable assertions when debug mode is enabled
+#ifdef _CCCL_ENABLE_DEBUG_MODE
+#  ifndef CCCL_ENABLE_ASSERTIONS
+#    define CCCL_ENABLE_ASSERTIONS
+#  endif // !CCCL_ENABLE_ASSERTIONS
+#endif // _CCCL_ENABLE_DEBUG_MODE
+
+//! Ensure that we switch on host assertions when all assertions are enabled
+#ifndef CCCL_ENABLE_HOST_ASSERTIONS
+#  ifdef CCCL_ENABLE_ASSERTIONS
+#    define CCCL_ENABLE_HOST_ASSERTIONS
+#  endif // CCCL_ENABLE_ASSERTIONS
+#endif // !CCCL_ENABLE_HOST_ASSERTIONS
+
+//! Ensure that we switch on device assertions when all assertions are enabled
+#ifndef CCCL_ENABLE_DEVICE_ASSERTIONS
+#  ifdef CCCL_ENABLE_ASSERTIONS
+#    define CCCL_ENABLE_DEVICE_ASSERTIONS
+#  endif // CCCL_ENABLE_ASSERTIONS
+#endif // !CCCL_ENABLE_DEVICE_ASSERTIONS
+
+//! Use internal nvcc implementation on device or the host library for clang-cuda
+#ifdef __CUDA_ARCH__
```

Comment:
This file, `cuda/std/__cccl/assert.h`, ignores NVC++ entirely, and won't work correctly for `nvc++ -stdpar=gpu`.  The file assumes that host and device code are compiled separately, and that `__CUDA_ARCH__` is defined when compiling device code and `__CUDA_ARCH__` is not defined when compiling host code.

`nvc++ -stdpar=gpu` compiles both host and device code in a single pass and does not define `__CUDA_ARCH__`.  It is probably necessary to give NVC++ special handling, separate from all the logic that is here.  `_CCCL_ASSERT_IMPL` would be defined as something that works in both host and device code with NVC++.  (`__assert_fail` might work in both host and device code with NVC++, but that should be verified.)

Reply:
I reworked the approach to use the host standard library on host and only use the special fallback for nvcc on device.

The `_CCCL_ASSERT_IMPL` now forwards to `_NV_IF_TARGET`